crates/rethnet_evm/src/state/overrides.rs

```diff
@@ -0,0 +1,178 @@
+use rethnet_eth::{account::KECCAK_EMPTY, Address, B256, U256};
+use revm::{
+    db::StateRef,
+    primitives::{AccountInfo, Bytecode, HashMap},
+};
+
+/// Type representing either a diff or full set of overrides for storage information.
+#[derive(Clone, Debug)]
+pub enum StorageOverride {
+    /// A diff of storage overrides.
+    Diff(HashMap<U256, U256>),
+    /// A full set of storage overrides.
+    Full(HashMap<U256, U256>),
+}
+
+/// Values for overriding account information.
+#[derive(Clone, Debug)]
+pub struct AccountOverride {
+    /// Account balance override.
+    pub balance: Option<U256>,
+    /// Account nonce override.
+    pub nonce: Option<u64>,
+    /// Account code override.
+    pub code: Option<Bytecode>,
+    /// Account storage override.
+    pub storage: Option<StorageOverride>,
+}
+
+impl AccountOverride {
+    /// Overrides the provided original account information.
+    pub fn override_info(&self, original: Option<AccountInfo>) -> Option<AccountInfo> {
+        let has_override = self.balance.is_some() | self.nonce.is_some() | self.code.is_some();
+
+        if !has_override {
+            return original;
+        }
+
+        let mut account_info = original.unwrap_or_default();
+        if let Some(balance) = &self.balance {
+            account_info.balance = *balance;
+        }
+
+        if let Some(nonce) = &self.nonce {
+            account_info.nonce = *nonce;
+        }
+
+        if let Some(code) = &self.code {
+            let code_hash = code.hash_slow();
+            if code_hash == KECCAK_EMPTY {
+                account_info.code = None;
+                account_info.code_hash = KECCAK_EMPTY;
+            } else {
+                account_info.code = Some(code.clone());
+                account_info.code_hash = code_hash;
+            }
+        }
+
+        Some(account_info)
+    }
+}
+
+/// A set of overrides for state information.
+#[derive(Clone, Debug, Default)]
+pub struct StateOverrides {
+    account_overrides: HashMap<Address, AccountOverride>,
+    code_by_hash_overrides: HashMap<B256, Bytecode>,
+}
+
+impl StateOverrides {
+    /// Constructs a new set of state overrides.
+    pub fn new(mut account_overrides: HashMap<Address, AccountOverride>) -> Self {
+        let code_by_hash_overrides = account_overrides
+            .values_mut()
+            .filter_map(|account_override| {
+                if let Some(code) = &mut account_override.code {
+                    let code_hash = code.hash_slow();
+
+                    Some((code_hash, code.clone()))
```

Comment:
We could potentially store a ref to the code to avoid [cloning it twice.](https://github.com/NomicFoundation/hardhat/pull/4463/files#diff-1e8df9b3c72556c829cc01444e78dae3bb39c57699dd0528f9e3e6bd5c474cbaR140)

Reply:
Under the hood, the `Bytes` class will already do something like that, whenever possible.